### PR TITLE
Add python sdk

### PR DESCRIPTION
**Pull request by shahruk10**
- added python sdk which abstracts away raw protobuf function calls

- added tests for client and TLS authentication

- updated docs with examples

### CODE REVIEW

File: grpc/py-cubic/cubic/client.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                                when connecting to an instance of cubic server that is 
+                                using a self-signed certificate.
+
+            clientCertificate:  PEM certificate as bytes string presented by this Client when
+                                connecting to a server. Use this when setting up mutually 
+                                authenticated TLS. The clientKey must also be provided.
+
+            clientKey:  PEM key as byte string presented by this Client when
+                        connecting to a server. Use this when setting up mutually 
+                        authenticated TLS. The clientKey must also be provided.
```
Comment (jhollowayj) [inline grpc/py-cubic/cubic/client.py]: The clientKey must also be provided.

`clientKey` or `clientCertificate`?

File: grpc/py-cubic/cubic/client.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        self.secure = secure
+        self.bufferSize = bufferSize
+        
+        if self.bufferSize <= 0:
+            raise ValueError('buffer size must be greater than 0')
+
+        if secure:
+            # using a TLS endpoint with optional certificates for mutual authentication
+            if clientCertificate is not None and clientKey is None:
+                raise ValueError("client key has not been provided")
```
Comment (jhollowayj) [inline grpc/py-cubic/cubic/client.py]: Maybe this is petty: change this to `raise ValueError("client key must also be provided")`?  Same would go for the other error.

File: grpc/py-cubic/cubic/client.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            self._channel = grpc.insecure_channel(serverAddress)
+
+        self._client = CubicStub(self._channel)
+
+    def __del__(self):
+        """ Closes and cleans up after Client. """
+        try:
+            self._channel.close()
+        except AttributeError:
+            # client wasn't
```
Comment (jhollowayj) [inline grpc/py-cubic/cubic/client.py]: client wasn't what?

File: grpc/py-cubic/cubic/test_client.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+from concurrent import futures
+
+from client import Client, RecognitionAudio, RecognitionConfig
+import cubic_pb2
+import cubic_pb2_grpc
+
+expectedResponses = {}
+expectedResponses['Version'] = cubic_pb2.VersionResponse(cubic='2727', server='v.27.0')
+expectedResponses['ListModels'] = cubic_pb2.ListModelsResponse(models=[cubic_pb2.Model(id="1"), cubic_pb2.Model(id="2")])
+expectedResponses['Recognize'] = cubic_pb2.RecognitionResponse(results=[cubic_pb2.RecognitionResult(alternatives=[cubic_pb2.RecognitionAlternative(transcript="This is a test")])])
```
Comment (jhollowayj) [inline grpc/py-cubic/cubic/test_client.py]: This line feels a little long.  Is there an easy way to to multi-line definitions?  I know I had a hard time in go deciding on the best format, so I understand if you say this is the best way for python

File: grpc/py-cubic/cubic/client.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+                    and sent to cubic server sequentially. The size of each chunk is equal to the 
+                    buffer size set for the client.
+        """
+
+        stream = _audioStreamer(cfg, audio, self.bufferSize)
+        for resp in self._client.StreamingRecognize(stream):
+            yield resp
+        
+def _audioStreamer(cfg, audio, bufferSize):
+    """ A generator that streams audio data packaged for streaming recognition.
```
Comment (jhollowayj) [inline grpc/py-cubic/cubic/client.py]: Thanks for using generators.  :+1:

File: docs-src/content/getting-started/streaming.md [hunk 6]
```diff
@@ -69,4 +69,37 @@ func main() {
...
+model = modelResp.models[0]
+
+cfg = cubic.RecognitionConfig(
+    model_id = model.Id
+)
+
+# audio file
+audio = open('test.raw', 'rb')
+
+# send streaming request to cubic and
```
Comment (catherinecobalt) [inline docs-src/content/getting-started/streaming.md]: How would this change if you want to stream audio from a mic and do live recognition? From this example I can't see what I'd change.

File: docs-src/content/getting-started/streaming.md [hunk 7]
```diff
@@ -69,4 +69,37 @@ func main() {
...
+{{% tab "Python" %}}
+``` python
+import cubic
+
+serverAddress = '127.0.0.1:2727'
+
+client = cubic.Client(serverAddress)
+
+# use the first available model
+modelResp = client.ListModels()
```
Comment (catherinecobalt) [inline docs-src/content/getting-started/streaming.md]: I think it'd be helpful to have something here which prints out all of the models, so clients can quickly see how to navigate the modelResp object

File: docs-src/content/getting-started/streaming.md [hunk 8]
```diff
@@ -69,4 +69,37 @@ func main() {
...
+model = modelResp.models[0]
+
+cfg = cubic.RecognitionConfig(
```
Comment (catherinecobalt) [inline docs-src/content/getting-started/streaming.md]: Some extra config options in here might be helpful to show (I assume this is now where things like asking for the CN, turning on formatting etc are configured). 

In general, I'd suggest showing more (or better, all) options in the sample code assuming customers can remove them if they don't want them. But a customer working out how to add a new option will take them some effort.

File: docs-src/content/getting-started/installation.md [hunk 9]
```diff
@@ -14,3 +14,11 @@ import this package into your application:
 ``` go
 import "github.com/cobaltspeech/sdk-cubic/grpc/go-cubic"
 ```
+
+### Python
+The Python SDK depends on Python >= 3.5. You may use pip to perform a system-wide install, or use virtualenv for a local install.
```
Comment (catherinecobalt) [inline docs-src/content/getting-started/installation.md]: pip or pip3?

File: docs-src/content/getting-started/recognize.md [hunk 10]
```diff
@@ -15,6 +15,7 @@ recognition.
 <!--more-->
```
Comment (catherinecobalt) [inline docs-src/content/getting-started/recognize.md]: This comment is unrelated to your PR but If streaming recognition is the preferred method to use cubic, I'd put that page first above this one (I'm not sure how the ordering of the pages is configured). Then I'd have a preamble here which says that 'If your audio is short you may prefer this way...'.

**Comment by catherinecobalt** [comment]
Great to see the new Python SDK! I only looked over the documentation, but I added some comments on the points where I didn't have all the information when I setup cubic with the old(?) sdk-client-python.

File: grpc/Makefile [hunk 11]
```diff
@@ -1,13 +1,15 @@
...
 
-all: go docs
+all: go py docs
 
 GO_OUTDIR=go-cubic/cubicpb
+PY_OUTDIR=py-cubic/cubic
 DOCS_OUTDIR=../docs-src/content/protobuf
 
 go: ${GO_OUTDIR}/cubic.pb.go ${GO_OUTDIR}/gw/cubic.pb.gw.go go-test
+py: $(PY_OUTDIR)/cubic_pb2.py ${PY_OUTDIR}/cubic_pb2_grpc.py py-test
```
[the remark this answers is not in the document]
Comment (happyalu) [reply grpc/Makefile]: that extension is chosen by the gen tools. probably ok to leave it as it is.

File: docs-src/content/getting-started/streaming.md (hunk 8 above)
Comment (happyalu) [reply docs-src/content/getting-started/streaming.md]: Good idea. We could add a section in the docs (across languages) to show how to configure a few useful things, such as these.

File: docs-src/content/getting-started/streaming.md (hunk 6 above)
Comment (happyalu) [reply docs-src/content/getting-started/streaming.md]: I think we should document that client.StreamingRecognize takes any object that has the read() method. Perhaps have commented-out line or two that uses microphone input.

File: grpc/Makefile [hunk 12]
```diff
@@ -34,10 +36,22 @@ ${GO_OUTDIR}/gw/cubic.pb.gw.go: cubic.proto
...
+${PY_OUTDIR}/cubic_pb2_grpc.py: cubic.proto
+	mkdir -p ${PY_OUTDIR}
+	python -m grpc_tools.protoc -I . -I ${GOPATH}/src/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis --grpc_python_out=${PY_OUTDIR} $<
+
 go-test:
 	cd go-cubic && go test
 	cd go-cubic/cubicpb/gw && go test
 
+py-test:
+	cd py-cubic/cubic && python test_client.py
```
[the remark this answers is not in the document]
Comment (happyalu) [reply grpc/Makefile]: Make executes each line in a new shell, so this should be ok.

File: grpc/py-cubic/cubic/client.py [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+sys.path.append(os.path.join(os.path.dirname(os.path.realpath(__file__))))
+from cubic_pb2_grpc import CubicStub
+from cubic_pb2 import RecognitionConfig, RecognitionAudio
+from cubic_pb2 import ListModelsRequest, RecognizeRequest, StreamingRecognizeRequest
+
+class Client(object):
+    """ A client for interacting with Cobalt's Cubic GRPC API."""
+
+    def __init__(self, serverAddress, secure=True,
```
Comment (happyalu) [inline grpc/py-cubic/cubic/client.py]: change "secure=True" to "insecure=False".

File: grpc/py-cubic/cubic/client.py [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+    def Recognize(self, cfg, audio):
+        """ Performs synchronous speech recognition: receive results after all audio
+        has been sent and processed. It is expected that this request be typically
+        used for short audio content: less than a minute long.  For longer content,
+        the `StreamingRecognize` method should be preferred.
+
+        Args:
+            cfg: RecognitionConfig object containing the model ID, timeout, audio encoding etc.
+
+            audio: a binary stream of data to send to cubic.
```
Comment (happyalu) [inline grpc/py-cubic/cubic/client.py]: probably clarify that this is some object that implements the read() method. Otherwise clients might attempt to use a bytestring/etc.

File: grpc/py-cubic/cubic/client.py [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+
+            audio: a binary stream of data to send to cubic. 
+        """
+
+        rcgAudio = RecognitionAudio(data=audio.read())
+        return self._client.Recognize(RecognizeRequest(config=cfg, audio=rcgAudio))
+
+    def StreamingRecognize(self, cfg, audio):
+        """ Performs bidirectional streaming speech recognition. Receive results while
+            sending audio. This method is only available via GRPC and not via
```
Comment (happyalu) [inline grpc/py-cubic/cubic/client.py]: There's no need to mention HTTP+JSON here. This is the GRPC API SDK.

File: grpc/py-cubic/setup.py [hunk 16]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env python
+from setuptools import setup
+
+setup(
+    name='cobalt-cubic',
+    python_requires='>=3.5.0',
+    version='2.0.0',
```
Comment (happyalu) [inline grpc/py-cubic/setup.py]: This version should be set to the same value as the [go sdk](https://github.com/cobaltspeech/sdk-cubic/blob/master/grpc/go-cubic/cubicpb/gw/go.mod#L6) and the [README](https://github.com/cobaltspeech/sdk-cubic/blob/master/README.md#tagging-new-versions) needs to be changed to also update this value before tagging a release.

File: grpc/py-cubic/setup.py [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+    description='This client library is designed to support the Cobalt API for speech recognition with Cubic',
+    author='Cobalt Speech and Language Inc.',
+    maintainer_email='tech@cobaltspeech.com',
+    url='https://github.com/cobaltspeech/sdk-cubic',
```
Comment (happyalu) [inline grpc/py-cubic/setup.py]: should this URL be the github URL, or the docs url (https://cobaltspeech.github.io/sdk-cubic) ?

File: docs-src/content/getting-started/streaming.md [hunk 18]
```diff
@@ -69,4 +69,37 @@ func main() {
...
+
+# send streaming request to cubic and 
+# print out results as they come in
+for resp in client.StreamingRecognize(cfg, audio):
+    for result in resp.results:
+        print(result.alternatives[0].transcript)
+        if result.is_partial:
+            print("... (partial)")
+        else:
+            print("   (final)")
```
[the remark this answers is not in the document]
Comment (shahruk10) [reply docs-src/content/getting-started/streaming.md]: yup this looks better than printing partial results on different lines, will change accordinly

File: grpc/py-cubic/cubic/client.py (hunk 1 above)
Comment (shahruk10) [reply grpc/py-cubic/cubic/client.py]: `clientCertificate`, will change

File: grpc/py-cubic/cubic/client.py (hunk 3 above)
Comment (shahruk10) [reply grpc/py-cubic/cubic/client.py]: oops, was supposed to be something along the lines of ` # client was't fully instantiated, no channel to close `

File: grpc/py-cubic/cubic/test_client.py (hunk 4 above)
Comment (shahruk10) [reply grpc/py-cubic/cubic/test_client.py]: true, it is long, I did break it into multiple lines but it looked ugly in my eyes. I will try and see if I can make it look cleaner

File: docs-src/content/getting-started/streaming.md (hunk 7 above)
Comment (shahruk10) [reply docs-src/content/getting-started/streaming.md]: good point, will do

File: docs-src/content/getting-started/installation.md (hunk 9 above)
Comment (shahruk10) [reply docs-src/content/getting-started/installation.md]: I think generally `pip` is soft linked to the appropriate version corresponding to the current environment's python version, so just `pip` should work

File: grpc/py-cubic/cubic/client.py (hunk 14 above)
Comment (shahruk10) [reply grpc/py-cubic/cubic/client.py]: will do; should we also document each function in the actual sdk documentation instead of only the code files ?

File: grpc/py-cubic/setup.py (hunk 17 above)
Comment (shahruk10) [reply grpc/py-cubic/setup.py]: I dunno, both the luna-sdk and the whole cubic-sdk-python had the github url

File: grpc/py-cubic/cubic/client.py [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import sys
+import grpc
+
+from google.protobuf.empty_pb2 import Empty
+
+sys.path.append(os.path.join(os.path.dirname(os.path.realpath(__file__))))
```
[the remark this answers is not in the document]
Comment (shahruk10) [reply grpc/py-cubic/cubic/client.py]: It's a long winded way of adding a directory pointed to by `__file__` to the import path.
`__file__` = pathname of the file from which the module was loaded.

It's needed so that the protoc generated files can find each other

File: grpc/py-cubic/cubic/client.py (hunk 14 above)
Comment (happyalu) [reply grpc/py-cubic/cubic/client.py]: Maybe later.

File: grpc/py-cubic/setup.py (hunk 17 above)
Comment (happyalu) [reply grpc/py-cubic/setup.py]: Let's use the docs url. The repo url as documentation of the package can be confusing at first glance since the repo is not just the python package

File: docs-src/content/getting-started/streaming.md (hunk 8 above)
Comment (shahruk10) [reply docs-src/content/getting-started/streaming.md]: will do this in a separate PR, resolving for now

File: docs-src/content/getting-started/streaming.md [hunk 20]
```diff
@@ -2,13 +2,16 @@
...
 weight: 30
 ---
-The following example shows how to transcribe an audio file using Cubic's
-Streaming Recognize Request. It is assumed that the audio file contains raw
-samples, PCM16SLE like Cubic expects.  We will query the server for available
-models and use the first model to transcribe this speech.
+The following examples show how to transcribe audio streams using Cubic's
+Streaming Recognize Request. The stream can come from a file on disk or be
+directly from a microphone in real time. It is assumed that the audio stream 
+contains raw samples, PCM16SLE like Cubic expects. We will query the server
```
Comment (happyalu) [inline docs-src/content/getting-started/streaming.md]: Since we support more formats now, should we perhaps use WAV as the example, since it will be a more common use case ?

File: docs-src/content/getting-started/streaming.md [hunk 21]
```diff
@@ -69,4 +72,111 @@ func main() {
...
+		else:
+			print("\r{0}".format(result.alternatives[0].transcript), end="\n")
+
+```
+{{% /tab %}}
+
+{{%/tabs %}}
+
+
+### Streaming from microphone
```
Comment (happyalu) [inline docs-src/content/getting-started/streaming.md]: Was this gonig to be a separate PR?

Comment (shahruk10) [reply docs-src/content/getting-started/streaming.md]: Removed the microphone example; will put it up together with a parallel example for the go client in a separate PR
